parsers/ZA.py

```diff
@@ -0,0 +1,961 @@
+    column_mapping = {
+        0: "coal",  # Thermal_Gen_Excl_Pumping_and_SCO
+        1: "unknown",  # Eskom_OCGT_SCO_Pumping             changed to unknown since negative oil is not possible
+        2: "unknown",  # Eskom_Gas_SCO_Pumping              changed to unknown since negative gas is not possible
+        3: "hydro",  # Hydro_Water_SCO_Pumping
+        4: "hydro",  # Pumped_Water_SCO_Pumping
+        5: "unknown",  # Thermal_Generation                 sum of 0, 1, 2, 3, 4. This value is not used.
+        6: "nuclear",  # Nuclear_Generation
+        7: "unknown",  # International_Imports
+        8: "oil",  # Eskom_OCGT_Generation
+        9: "gas",  # Eskom_Gas_Generation
+        10: "oil",  # Dispatchable_IPP_OCGT
+        11: "hydro",  # Hydro_Water_Generation
+        12: "hydro",  # Pumped_Water_Generation
+        13: "unknown",  # IOS_Excl_ILS_and_MLR              Interruption of Supply. Not used.
+        14: "unknown",  # ILS_Usage                         Interruptible Load Shed = companies paid not to consume electricity. Not used.
+        15: "unknown",  # Manual_Load_Reduction_MLR         MLS = forced load shedding. Not used.
+        16: "wind",  # Wind
+        17: "solar",  # PV
+        18: "solar",  # CSP
+        19: "unknown",  # Other_RE
+    }
+
+    all_data = []
+
+    for i, (datetime_key, value) in enumerate(merged_csv_bi_data.items()):
+
+        # The production values (MW) should never be negative. Use None, or omit the key if # a specific production mode is not known.
+        # ---
+        # storage values can be both positive (when storing energy) or negative (when the
+        # storage is discharged).
+
+        data = {
+            "zoneKey": zone_key,
+            "datetime": datetime_key.datetime,
+            "production": {
+                "biomass": None,
+                "coal": 0.0,
+                "gas": 0.0,
+                "nuclear": 0.0,
+                "oil": 0.0,
+                "solar": 0.0,
+                "wind": 0.0,
+                "geothermal": None,
+                "unknown": 0.0,
+            },
+            "storage": {"hydro": 0.0},
+            "source": " https://www.eskom.co.za",
+        }
+
+        cleansed_csv_values = value["cleansed_csv"].split(",")
+
+        storage_inversion_idcs = [3, 4, 11, 12]
+        production_idcs = [0, 1, 2, 6, 7, 8, 9, 10, 16, 17, 18, 19]
+
+        for j, cleansed_csv_value in enumerate(cleansed_csv_values):
+            if j in storage_inversion_idcs:
+                data["storage"][column_mapping[j]] = round(
+                    data["storage"][column_mapping[j]] + float(cleansed_csv_value) * -1,
+                    13,
+                )
+            elif j in production_idcs:
+                data["production"][column_mapping[j]] = round(
+                    data["production"][column_mapping[j]] + float(cleansed_csv_value),
+                    13,
+                )
```

Comment:
```suggestion
        for j, cleansed_csv_value in enumerate(cleansed_csv_values):
            if j in storage_inversion_idcs:
                data["storage"][column_mapping[j]] = round(
                    data["storage"][column_mapping[j]] + float(cleansed_csv_value) * -1,
                    2,
                )
            elif j in production_idcs:
                data["production"][column_mapping[j]] = round(
                    data["production"][column_mapping[j]] + float(cleansed_csv_value),
                    2,
                )
```
2 decimals should be enough but at most it should be 4, we don't need 13 decimals or microwatt scale (if my math checks out).

Reply:
Done in commit 201c005b280d56fa05a757e8b1d3963e138f05a3

Reply:
> 2 decimals should be enough but at most it should be 4, we don't need 13 decimals or microwatt scale (if my math checks out).

This is only done to eliminate floating point errors,  it should not result in numbers with 13 actual decimal places unless those are actually in the data source. 13 is due to the size of Python's floats, 14 doesn't get rid of the floating point errors.

Anyway, I don't think  rounding it that much is a good idea. On a low production grid it'll actually make a significant difference (e.g. solar on the [Faroe Islands](https://app.electricitymaps.com/zone/FO?wind=true) is less than 100kW, so you see 10kW, 20kW, 70kW - more precision would be actually welcome there.

Furthermore, I think it would be better to leave the precision of the stored data to ElectricytMap's back-end systems, not the parser. That makes it easier to increase the precision if desired, without having to edit multiple individual parsers.

Reply:
I confirm, in general the closer the data we receive from the parsers is from the data source the best it is so we can then have the freedom to design a consistent set of operations in the backend